### ISSUE TITLE
convergeP

### ISSUE THREAD

**Issue by asaf-romano** [ramda/ramda#1485]
Let's assume all of the following are unary functions taking the same user identifier: ``getUserName``, ``getDisplayName``, ``getEmailAddress``.

Something like getUserDetails can then be defined like this:
```js
  let getUserDetails = converge(zipObj(["id", "userName", "displayName", "emailAddress"]), [identity, getUserName, getDisplayName, getEmailAddress]);
```

This works.. until all those get* functions become promise* functions. Then the way out looks like this:
```js
let promiseUserDetails = pipeP(
  // ::Promise.all is sugar for Promise.all.bind(Promise) - Promise.all isn't bound in the default implementation
  converge(unapply(::Promise.all), [identity, promiseUserName...]),
  zipObj([....])
);
```
Back when ``pipeP`` could take a synchronous function as its first function (see #1314), this also worked:
```js
let promiseUserDetails = pipeP(
  converge(Array.from, [identity, promiseUserName...]),
  ::Promise.all,
  zipObj([....])
);
```

With ``convergeP``, the above code could be written like this:
```js
let promiseUserDetails = convergeP(unapply(zipObj([...]), [identity, promiseUserName...]);
```
...which is somewhat more expressive.

There are three problems I see:
* simple converge is under attack (#1447)
* Ramda would need to simulate ``Promise``.all (or assume that if ``pipeP``/``composeP``/``convergeP`` are used, ``Promise`` is there).
* the converge function being ``unapply``ed would be the norm.

**Comment by scott-christopher** [comment]
Perhaps an alternative approach to support this would be to provide dispatching for `Promise` instances in `map` and `ap`, which would allow `lift` to be used for Promises and the existing `converge` could then be utilised.
e.g. something like the following:
```js
var liftedFn = lift(zipObj(["id", "userName", "displayName", "emailAddress"]));
var promiseUserDetails = converge(liftedFn, [Promise.resolve, getUserName, getDisplayName, getEmailAddress]);
```

**Comment by asaf-romano** [comment]
Yes, please :)

I thought it was a deliberate decision not to treat Promises as if they were mondas (perhaps as a way of encouraging users to use Futures instead), but I think the pros of doing so overcome the shortcomings, esp. now that ES6 and ES7 (async functions) made Promises first class citizens in JS.

**Comment by scott-christopher** [comment]
The downside is that it's not so straightforward to obey the functor composition law for Promises – `map(compose(f, g))` = `compose(map(f), map(g))` – If we proxy straight through to `Promise::then` and `g` returns another Promise, `Promise::then` will behave like `chain` rather than `map`.

Not sure whether this is a deal-breaker or just in need of some disclaimers in documentation.

**Comment by asaf-romano** [comment]
Closing this as I don't see any of the current owners willing to extend promises support ;)

**Comment by CrossEye** [comment]
For whatever reason, I never even saw this issue.

But you're probably right.  There's not much love for Promises among the core contributors.

**Comment by tencircles** [comment]
Would something like this work?

````
// convergeP :: (((a, b, c) -> d), [(e -> Promise)], f) -> Promise(d)
const convergeP = curry((f, arg_fs, input) => {
    return Promise.all(map(arg_f => arg_f(input), arg_fs))
        .then(args => f(...args));
});
````

**Comment by evan-king** [comment]
@tencircles I think so, though the naming is a bit hard to follow.  I think you're losing the ability to pass multiple arguments though.  Here's a pure ES5 version, which also handles a mix of sync/async functions:
```js
function convergeP = curry(function(merger, fns, arg) {
    var input = Promise.resolve(arg);
    return Promise.all(map(input.then, fns)).then(apply(merger));
});
```
When I find need for a `convergeP`, it almost always involves merging sync and async dataflows.  I actually more frequently need a way to merge multiple non-function possibly-async inputs, which I do with this `mergeP`, which I then use to help define `convergeP`:

```js
var mergeP = curry(function(merger, inputs) {
    return composeP(apply(merger), bind(Promise.all, Promise))(inputs);
});

//  Elegantly terse, but loses support for multiple arguments
// var convergeP = curry(function(merger, fns, arg) {
//     return mergeP(merger, map(applyTo(arg), fns));
// });

var applyThese = curry(function(args, fn) {
    return fn.apply(fn, args);
});

var convergeP = curry(function(merger, fns) {
    return function converger() {
        var args = Array.prototype.slice.call(arguments);
        return mergeP(merger, map(applyThese(args), fns));
    }
});
```

In practice though, I find myself needing the flexibility of my `mergeP` and unable to find a point-free `convergeP` approach.

**Comment by evan-king** [comment]
I don't think an improved implementation/version addresses the original reason for closing.  I shared it for the benefit of others willing to use a non-core implementation.  I think #2565 is the nail in the coffin for `convergeP`, and may lead to alternative approaches anyway.

**Comment by semmel** [comment]
For fun with Ramda `0.26`
```javascript
var mergeP = merger => R.compose(R.then(R.apply(merger)), Promise.all.bind(Promise));
var applyThese = R.flip(R.apply);
```

But the fun will be all too fleeting because, just being implemented, `R.then` is already threatened to [get axed next time](#2751). :disappointed:

**Comment by evan-king** [comment]
I share your disappointment over `R.then`.  It was the knight in shining armor come to save us all from the Ramda community's hatred of promises.

No hard feelings if the moderators remove this comment for advertising, but these days I use a library of my own making to handle a lot of promise synchronization tasks: [promise-apply-spec](https://github.com/evan-king/promise-apply-spec).  It doesn't really cover the loss of `R.then` but does allow building more robust `spec` objects that can sometimes assemble all the same logic which would otherwise be composed.  In truth I'd rather use it as fallback than go-to approach, but one major advantage is better insight into failure conditions (captures _all_ rejections).

**Comment by CrossEye** [comment]
I am going to submit a PR to rename `then` instead of removing it.  I think it's worth having.  I just am in the middle of a major bit of Ramda documentation work, and am trying not to get distracted from it.
